src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultDocumentContextFactory.cs

```diff
@@ -0,0 +1,83 @@
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal class DefaultDocumentContextFactory : DocumentContextFactory
+    {
+        private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+        private readonly DocumentResolver _documentResolver;
+        private readonly DocumentVersionCache _documentVersionCache;
+        private readonly ILogger<DefaultDocumentContextFactory> _logger;
+
+        public DefaultDocumentContextFactory(
+            ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+            DocumentResolver documentResolver,
+            DocumentVersionCache documentVersionCache,
+            ILoggerFactory loggerFactory)
+        {
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _documentVersionCache = documentVersionCache;
+            _logger = loggerFactory.CreateLogger<DefaultDocumentContextFactory>();
+        }
+
+        public override async Task<DocumentContext?> TryCreateAsync(Uri documentUri, CancellationToken cancellationToken)
+        {
+            var filePath = documentUri.GetAbsoluteOrUNCPath();
+            var documentAndVersion = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                if (_documentResolver.TryResolveDocument(filePath, out var documentSnapshot))
+                {
+                    if (_documentVersionCache.TryGetDocumentVersion(documentSnapshot, out var version))
+                    {
+                        return new DocumentSnapshotAndVersion(documentSnapshot, version.Value);
+                    }
+                }
+
+                return null;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentAndVersion is null)
+            {
+                return null;
+            }
+
+            var (documentSnapshot, version) = documentAndVersion;
+            if (documentSnapshot is null)
+            {
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                _logger.LogWarning($"Failed to retrieve generated output for document {documentUri}. It is unsupported.");
+                return null;
+            }
+
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
```

Comment:
Should this be done lazily in the context instead? Seems to add no value here...

Reply:
It could; however, that'd then require us to have an async method on the "documentcontext" that calculates this. I was hoping to keep the contract as synchronous as possible. Also, I thinkkkkkk this is free once code document calculation is done; however, definitely worth checking because if it's not then that'd definitely be another reason to lazily do it.

Reply:
Thats fair, but if this becomes the base for methods, or extension methods, that get syntax nodes etc. I wouldn't be surprised if that doesn't hold for very long 😛

Reply:
Ya I feel like if we move this to be an asynchronous call we'd probably want to do the same for `GetCodeDocumentAsync` above.  Hmmmmmmmmmmmm and I guessss those methods could potentially cache the result resulting in the same "execute once" behavior.

Reply:
@ryanbrandenburg also had similar feedback. Ok 2 vs. 1 😄, I'll see what it means to migrate!

Reply:
Actually it ended up being totally reasonable and I could maintain the caching perf wins: https://github.com/dotnet/razor-tooling/pull/6435/commits/6ae11db7277d6f8fcc06f15b929c9fa1b92d60a9